src/devices/Ili934x/Chsc6440.cs

```diff
@@ -0,0 +1,410 @@
+                }
+
+                _gpioController.OpenPin(_interruptPin, PinMode.InputPullUp);
+                _gpioController.RegisterCallbackForPinValueChangedEvent(_interruptPin, PinEventTypes.Rising | PinEventTypes.Falling, OnInterrupt);
+            }
+        }
+
+        /// <summary>
+        /// Size of the screen
+        /// </summary>
+        public Size ScreenSize { get; }
+
+        /// <summary>
+        /// Sets the background thread update interval. Low values can impact performance, but increase the responsiveness.
+        /// </summary>
+        public TimeSpan UpdateInterval
+        {
+            get
+            {
+                return _interval;
+            }
+            set
+            {
+                _interval = value;
+            }
+        }
+
+        /// <summary>
+        /// The size of the rectangle that is considered a "touch". When the position changes more than this, it is considered a drag.
+        /// </summary>
+        public Size TouchSize
+        {
+            get;
+            set;
+        }
+
+        private void OnInterrupt(object sender, PinValueChangedEventArgs pinValueChangedEventArgs)
+        {
+            // The pin is low as long as the screen is being touched.
+            // So we still have to poll when the user drags
+            if (pinValueChangedEventArgs.PinNumber != _interruptPin)
+            {
+                return;
+            }
+
+            _isPressed = pinValueChangedEventArgs.ChangeType == PinEventTypes.Falling;
+            _updateEvent.Set();
+        }
+
+        /// <summary>
+        /// Returns true if the interrupt pin is set, meaning something is touching the display
+        /// </summary>
+        /// <returns>True if something presses the display, false if not. This queries the interrupt pin if available. Otherwise, an I2C request to the controller is required.</returns>
+        public bool IsPressed()
+        {
+            if (_gpioController != null)
+            {
+                return _isPressed;
+            }
+
+            // Need to query the device instead
+            Span<byte> register = stackalloc byte[1]
+            {
+                0x02
+            };
+
+            Span<byte> result = stackalloc byte[1];
+
+            _i2c.WriteRead(register, result);
+
+            return result[0] != 0;
+        }
+
+        private void ReadData()
+        {
+            lock (_lock)
+            {
+                // true if real read, not a "come back later"
+                _wasRead = false;
+
+                Span<Point> p = stackalloc Point[2];
+                p[0] = default;
+                p[1] = default;
+                byte pts = 0;
+                int p0f = 0;
+                Span<byte> register = stackalloc byte[1]
+                {
+                    0x02
+                };
+
+                if (IsPressed())
+                {
+                    Span<byte> data = stackalloc byte[11];
```

Comment:
(non-blocking) could some structure be used here instead? the bit twiddling below looks a bit confusing

Reply:
I agree, but this is just the internal decoding of a binary I2C reply, as we do it in many other bindings as well. This unpacks the bits into structures.